staging/src/k8s.io/controller-manager/pkg/clientbuilder/client_builder_dynamic.go

```diff
@@ -156,6 +162,37 @@ func (t *DynamicControllerClientBuilder) ClientOrDie(name string) clientset.Inte
 	return client
 }
 
+type resettableTokenSource interface {
+	oauth2.TokenSource
+	ResetToken()
+}
+
+type resettableTokenSourceTransport struct {
+	base http.RoundTripper
+	src  resettableTokenSource
+}
+
+func (rt *resettableTokenSourceTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	resp, err := rt.base.RoundTrip(req)
+	if apierrors.IsUnauthorized(err) {
+		rt.src.ResetToken()
+		return rt.base.RoundTrip(req)
+	}
+	return resp, err
+}
```

Comment:
this seems useful enough to have in `staging/src/k8s.io/client-go/transport/token_source.go` or to build into `tokenSourceTransport`, rather than here

* `NewCachedTokenSource` could return ResettableTokenSource
* a new `ResettableTokenSourceWrapTransport` constructor could populate an optional resetToken field in `tokenSourceTransport`

That would let you avoid re-implementing round-trip and cancelrequest

Reply:
done